src/test/sql/sql_pipeline_test.cpp

```diff
@@ -0,0 +1,385 @@
+  }
+
+  bool _contains_validate(const std::shared_ptr<AbstractLQPNode>& root_node) {
+    auto node = root_node;
+    while (node != nullptr) {
+      if (node->type() == LQPNodeType::Validate) return true;
+      node = node->left_child();
+    }
+    return false;
+  }
+
+  std::shared_ptr<Table> _table_a;
+  std::shared_ptr<Table> _table_b;
+  std::shared_ptr<Table> _join_result;
+
+  const std::string _select_query_a = "SELECT * FROM table_a";
+  const std::string _invalid_sql = "SELECT FROM table_a";
+  const std::string _join_query =
+      "SELECT table_a.a, table_a.b, table_b.b AS bb FROM table_a JOIN table_b ON table_a.a = table_b.a WHERE table_a.a "
+      "> 1000";
+};
+
+TEST_F(SQLPipelineTest, SimpleCreation) {
+  SQLPipeline sql_pipeline(_select_query_a);
+
+  EXPECT_NE(sql_pipeline.transaction_context(), nullptr);
+  EXPECT_EQ(sql_pipeline.sql_string(), _select_query_a);
+}
+
+TEST_F(SQLPipelineTest, SimpleCreationWithoutMVCC) {
+  SQLPipeline sql_pipeline(_select_query_a, false);
+
+  EXPECT_EQ(sql_pipeline.transaction_context(), nullptr);
+}
+
+TEST_F(SQLPipelineTest, SimpleCreationWithCustomTransactionContext) {
+  auto context = TransactionManager::get().new_transaction_context();
+  SQLPipeline sql_pipeline(_select_query_a, context);
+
+  EXPECT_EQ(sql_pipeline.transaction_context().get(), context.get());
+  EXPECT_EQ(sql_pipeline.sql_string(), _select_query_a);
+}
+
+TEST_F(SQLPipelineTest, GetParsedSQL) {
+  SQLPipeline sql_pipeline(_select_query_a);
+  const auto& parsed_sql = sql_pipeline.get_parsed_sql();
+
+  EXPECT_TRUE(parsed_sql.isValid());
+
+  auto statements = parsed_sql.getStatements();
+
+  EXPECT_EQ(statements.size(), 1ul);
+  EXPECT_EQ(statements.at(0)->type(), hsql::StatementType::kStmtSelect);
+}
+
+TEST_F(SQLPipelineTest, GetParsedSQLTwice) {
+  SQLPipeline sql_pipeline(_select_query_a);
+  sql_pipeline.get_parsed_sql();
+  const auto& parsed_sql = sql_pipeline.get_parsed_sql();
+
+  EXPECT_TRUE(parsed_sql.isValid());
+
+  auto statements = parsed_sql.getStatements();
+
+  EXPECT_EQ(statements.size(), 1ul);
+  EXPECT_EQ(statements.at(0)->type(), hsql::StatementType::kStmtSelect);
+}
+
+TEST_F(SQLPipelineTest, GetParsedSQLInvalid) {
+  SQLPipeline sql_pipeline(_invalid_sql);
+  EXPECT_THROW(sql_pipeline.get_parsed_sql(), std::exception);
+}
+
+TEST_F(SQLPipelineTest, GetUnoptimizedLQP) {
+  SQLPipeline sql_pipeline(_select_query_a);
+
+  const auto& lqp_roots = sql_pipeline.get_unoptimized_logical_plan();
+
+  EXPECT_EQ(lqp_roots.size(), 1ul);
+  EXPECT_EQ(lqp_roots.at(0)->type(), LQPNodeType::Projection);
+}
+
+TEST_F(SQLPipelineTest, GetUnoptimizedLQPTwice) {
+  SQLPipeline sql_pipeline(_select_query_a);
+
+  sql_pipeline.get_unoptimized_logical_plan();
+  const auto& lqp_roots = sql_pipeline.get_unoptimized_logical_plan();
+
+  EXPECT_EQ(lqp_roots.size(), 1ul);
+  EXPECT_EQ(lqp_roots.at(0)->type(), LQPNodeType::Projection);
+}
+
+TEST_F(SQLPipelineTest, GetUnoptimizedLQPWithJoinFilter) {
```

Comment:
The SQLPipelineTest shouldn't depend on the structure of the LQP, if there is no other way to check whether this returns an unoptimized LQP, better not test it at all.

Reply:
Technically, there is no "unoptimized LQP". We can't tell if the optimizer ran on an LQP or not. Personally, I would be fine with this. Otherwise, we could also take a SQL query that should be optimized and check that unoptimized and optimized are different.